chia/util/keychain.py

```diff
@@ -132,6 +143,104 @@ def get_private_key_user(user: str, index: int) -> str:
+        # keys here.
+        try:
+            bytes_from_mnemonic(self.mnemonic_str())
+        except Exception as e:
+            raise KeychainKeyDataMissmatch("mnemonic") from e
+        if bytes_from_mnemonic(self.mnemonic_str()) != self.entropy:
+            raise KeychainKeyDataMissmatch("entropy")
+        if AugSchemeMPL.key_gen(mnemonic_to_seed(self.mnemonic_str())) != self.private_key:
+            raise KeychainKeyDataMissmatch("private_key")
+
+    @classmethod
+    def from_mnemonic(cls, mnemonic: str) -> KeyDataSecrets:
+        return cls(
+            mnemonic=mnemonic.split(),
+            entropy=bytes_from_mnemonic(mnemonic),
+            private_key=AugSchemeMPL.key_gen(mnemonic_to_seed(mnemonic)),
+        )
+
+    @classmethod
+    def from_entropy(cls, entropy: bytes) -> KeyDataSecrets:
+        return cls.from_mnemonic(bytes_to_mnemonic(entropy))
+
+    @classmethod
+    def generate(cls) -> KeyDataSecrets:
+        return cls.from_mnemonic(generate_mnemonic())
+
+    def mnemonic_str(self) -> str:
+        return " ".join(self.mnemonic)
+
+
+@final
+@streamable
+@dataclass(frozen=True)
+class KeyData(Streamable):
+    fingerprint: uint32
+    public_key: G1Element
+    secrets: Optional[KeyDataSecrets]
+
+    def __post_init__(self) -> None:
+        # This is redundant if `from_*` methods are used but its to make sure there can't be an `KeyData` instance with
+        # an attribute missmatch for calculated cached values. Should be ok since we don't handle a lot of keys here.
+        if self.secrets is not None and self.public_key != self.private_key.get_g1():
+            raise KeychainKeyDataMissmatch("public_key")
+        if self.public_key.get_fingerprint() != self.fingerprint:
+            raise KeychainKeyDataMissmatch("fingerprint")
+
+    @classmethod
+    def from_mnemonic(cls, mnemonic: str) -> KeyData:
+        private_key = AugSchemeMPL.key_gen(mnemonic_to_seed(mnemonic))
+        return cls(
+            fingerprint=private_key.get_g1().get_fingerprint(),
+            public_key=private_key.get_g1(),
+            secrets=KeyDataSecrets.from_mnemonic(mnemonic),
+        )
+
+    @classmethod
+    def from_entropy(cls, entropy: bytes) -> KeyData:
+        return cls.from_mnemonic(bytes_to_mnemonic(entropy))
+
+    @classmethod
+    def generate(cls) -> KeyData:
+        return cls.from_mnemonic(generate_mnemonic())
+
+    @property
+    def mnemonic(self) -> List[str]:
+        if self.secrets is None:
+            raise KeychainSecretsMissing()
+        return self.secrets.mnemonic
+
+    def mnemonic_str(self) -> str:
+        if self.secrets is None:
+            raise KeychainSecretsMissing()
+        return self.secrets.mnemonic_str()
+
+    @property
+    def entropy(self) -> bytes:
+        if self.secrets is None:
+            raise KeychainSecretsMissing()
+        return self.secrets.entropy
+
+    @property
+    def private_key(self) -> PrivateKey:
+        if self.secrets is None:
```

Comment:
Can this be moved into `__post_init__()` to avoid the duplicated check for each property/instance-method?

Reply:
Moving this check into `__post_init__` would make impossible to have an `KeyData` instance without secrets and would anyway also still require some assert then or so to make `mypy` happy.